src/Lorentz/Contracts/RegistryDAO.hs

```diff
@@ -0,0 +1,125 @@
+-- SPDX-FileCopyrightText: 2020 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+-- | A Registry DAO can be used to store key value pair `(k, v)`
+module Lorentz.Contracts.RegistryDAO
+  ( registryDaoContract
+  , config
+  )
+  where
+
+import Lorentz
+
+import qualified Lorentz.Contracts.BaseDAO as DAO
+import Lorentz.Contracts.RegistryDAO.Types
+import qualified Lorentz.Contracts.BaseDAO.Types as DAO
+import Universum hiding (drop, swap, (>>), some)
+import Util.Markdown
+
+-- | Proposal check: `s = size(pack(proposalMetadata))` `(a * s + b)`
+registryDaoProposalCheck
+  :: forall s k v.
+     ( NicePackedValue (RegistryDaoProposalMetadata k v)
+     )
+  => DAO.ProposeParams (RegistryDaoProposalMetadata k v) : s
+  :-> Bool : s
+registryDaoProposalCheck = do
+  let a :: Natural = 1; b :: Natural = 0
+  getField #ppProposalMetadata
+  packRaw;
+  size;
+  push a; mul
+  push b; add
+  toNamed #requireValue; swap
+  toFieldNamed #ppFrozenToken; swap
+  if #requireValue <=. #ppFrozenToken then
+    push True
+  else
+    push False
+
+-- |  slash_value ( c * frozen / d), c = 1, d = 1
+registryDaoRejectedProposalReturnValue
+  :: forall s k v. (IsoRegistryDaoProposalMetadata k v) => (DAO.Proposal (RegistryDaoProposalMetadata k v)) : s
+  :-> ("slash_amount" :! Natural) : s
+registryDaoRejectedProposalReturnValue = do
+  let c :: Natural = 1; d :: Natural = 1
+  toField #pProposerFrozenToken
+  push c; mul
+  push d; swap; ediv
+  ifSome car $
+    push (0 :: Natural)
+  toNamed #slash_amount
+
+-- | Update the propose changes
+decisionLambda
+  :: forall k v s store.
+  ( DAO.StorageC store (RegistryDaoContractExtra k v) (RegistryDaoProposalMetadata k v)
+  -- shared
+  , IsoRegistryDaoProposalMetadata k v
+  , NicePackedValue (RegistryDaoProposalMetadata k v)
+  )
+  =>  DAO.Proposal (RegistryDaoProposalMetadata k v) : store : s
+  :-> List Operation : store : s
+decisionLambda = do
+  toField #pMetadata
+  getField #pmAgoraPostId
+  swap; toField #pmDiff
+  iter $ do
+    stackType @(RegistryUpdate k v : Natural : store : s)
+    constructT @(RegistryEntry v)
+      ( fieldCtor $ getField #ruNewValue
+      , fieldCtor $ duupX @2
+      , fieldCtor $ now
+      )
+    stackType @(RegistryEntry v : RegistryUpdate k v : Natural : store : s)
+    dip $ dip swap
+    dip $ toField #ruKey
+    some; swap;
+    duupX @3; stToField #sExtra;
+    stackType @((RegistryDaoContractExtra k v) : k : Maybe (RegistryEntry v) : store : Natural : s)
+    toField #ceRegistry
+    stackType @(BigMap k (RegistryEntry v) : k : Maybe (RegistryEntry v) : store : Natural : s)
+    swap; dip swap; update -- Update `ceRegistry`
+    duupX @2; stToField #sExtra; swap; setField #ceRegistry -- -- Update `sExtra`
+    stSetField #sExtra
+    swap
```

Comment:
Minor suggestion: you could provide a shortcut access to the registry map via defining

```
instance StoreHasSubmap (Storage RegistryDaoContractExtra RegistryDaoProposalMetadata) "sRegistry" k (RegistryEntry v) where
  -- approximate implementation
  storeSubmapOps = composeStoreSubmapOps #ceRegistry storeFieldOps storeSubmapOps  
```

Then calling these getters/setters to update the registry map won't be necessary, you can just call `stInsert` on the entire storage.

Reply:
I'm not exactly sure how to use `composeStoreSubmapOps`. Currently this causes error: 
```
baseDAO    >     • Datatype DAO.Storage
baseDAO    >                  (RegistryDaoContractExtra k v)
baseDAO    >                  (RegistryDaoProposalMetadata k v) has no field "ceRegistry"
```

Reply:
Oh right, I forgot that our contract extra is part of the entire DAO storage.

Then I would define this instance not for `Storage ...` but for `RegistryDaoContractExtra`, and would use `stInsert` on the `RegistryDaoContractExtra`.

Reply:
Oh, that makes sense now. I added that, and update the code above to:

```haskell
..
          stackType @((RegistryDaoContractExtra k v) : k : Maybe (RegistryEntry v) : store : Natural : s)
          dig @2; dig @2; stUpdate #ceRegistry
          stSetField #sExtra
          swap
```

Reply:
Cool, thanks :+1: